### DISCUSSION TITLE
Documenting General Purpose Inputs and Outputs inside netbox

### DISCUSSION THREAD

**Discussion by HackXIt** [netbox-community/netbox#14120] [Ideas]
Greetings,

in my position as a test automation engineer for embedded systems, I'm working on a project, which involves the creation of a testing rack for a multitude of embedded products.
That project has already progressed into a stage, where we have started to build some test panels with proprietary devices on them.

The question of documentation for connections came up in the initial phase of building this first test panel.
I have used netbox in the past for IT racks in our manual testing facility, which worked greatly in documenting all sort of stuff revolving around IP/LAN.

I was wondering, is there any possibility of being able to document proprietary I/Os, in other words, electrical connections, that would make netbox a suitable option for documenting our whole test automation setup?

We use Raspberry Pi's connected to proprietary I/O boards that interact with our products on an electrical level.
A certain test panel might have over 250 I/O connections that all track various functionality.

It would be great to document all these connections and digital functions inside netbox.
But for that to be a possibility, I would need the ability to have more customization that would allow me to document those embedded systems.

Think of it as being an extended netbox, that is also able to document GPIO connections of a Raspberry Pi, with the termination having a certain functionality attached to it. Like this:
Raspberry Pi
- GPIO1 <-> LED input <-> Device A
- GPIO2 <-> Trigger button <-> Device A
- Raspberry Pi <-> GPIO3 <-> I²C Clock <-> ... <-> Raspberry Pi Pico
- Raspberry Pi <-> GPIO4 <-> I²C Data <-> ... <-> Raspberry Pi Pico

Can anyone help me out on this or has done something similar in the past?
Is this even a possibility in netbox?
Would it require coding a plugin, or would it require extending netbox functionality?

**Comment by HackXIt** [comment]
Anyone?

**Comment by candlerb** [comment]
Using the out-of-the-box functionality you could create each GPIO pin as an "interface" of type "Other", then you could run a "cable" between them.  This will let you track accurately which pin is connected to which other pin, which is probably what you need, but it doesn't model the physical reality well (e.g. it might be a single cable, like a ribbon cable, with multiple connected strands). 

You can also create interfaces on a module, to model things like a Pi Hat. 

Netbox does let you create a single "cable" with multiple terminations at each end, but it doesn't keep track of which termination at one end maps to which termination at the other end (indeed, there's no requirement for the number of terminations to match, as is the case for a QSFP breakout cable for example).

For MPO cables there's the frig for "multiple front-ports link to one rear-port", but that's not really useful for anything other than back-to-back patch panels with pass-through connections.

Making a whole new cable model as a plugin sounds to me like a major undertaking, and it's unclear to me how it would interact with Netbox's existing cabling model.

Perhaps what you could do is to create a "Bundle" plugin model, and then add a custom field on Cable pointing to Bundle, as a way of associating multiple cables together in one physical cable.  But note that it's still very difficult to connect a cable to a cable (e.g. for a soldered joint or a fibre splice); you'd need a dummy device with a front port and a rear port, as a joiner. And you can't have a cable which is connected at only one end.

**Reply by HackXIt** [reply]
I'll take a look at the interfaces then, it's a start at least. Since I'm also working on the server code of the raspberry pi, it would be a possibility of making it automatic, meaning I could build an integration for the net box API to add custom cables according to whatever I/Os are available and then all that's left for the human is to label them according to their specified function.
